pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +103,217 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	newMapiMachineSet.Spec.Template.Labels = util.MergeMaps(mapiMachineSet.Spec.Template.Labels, newMapiMachineSet.Spec.Template.Labels)
+
+	// Check if there are any changes after updating the spec
+	if !reflect.DeepEqual(newMapiMachineSet.Spec, mapiMachineSet.Spec) {
+		logger.Info("Updating MAPI machine set")
+
+		// The MachineSet that comes out of the conversion package is rejected by the APIServer (we don't set anything other than `.spec` currently).
+		mapiMachineSet.Spec = newMapiMachineSet.Spec
+
+		if err := r.Update(ctx, mapiMachineSet); err != nil {
+			logger.Error(err, "Failed to update MAPI machine set")
+
+			r.updateSynchronizedConditionWithPatch(
+				ctx, mapiMachineSet, corev1.ConditionFalse,
+				"FailedToUpdateMAPIMachineSet", err.Error(),
+			)
+
+			return ctrl.Result{}, fmt.Errorf("failed to update MAPI machine set: %w", err)
+		}
+
+		logger.Info("Successfully updated MAPI machine set")
+	} else {
+		logger.Info("No changes detected in MAPI machine set")
+	}
+
+	r.updateSynchronizedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionTrue,
+		consts.ReasonResourceSynchronized, "Successfully synchronized CAPI MachineSet to MAPI",
+		capiMachineSet.Generation)
+
 	return ctrl.Result{}, nil
 }
 
+// updateSynchronizedConditionWithPatch updates the synchronized condition
+// using a server side apply patch. We do this to force ownership of the
+// 'Synchronized' condition and 'SynchronizedGeneration'.
+func (r *MachineSetSyncReconciler) updateSynchronizedConditionWithPatch(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, status corev1.ConditionStatus, reason, message string, generation ...int64) {
+	logger := log.FromContext(ctx).WithValues("function", "updateSynchronizedConditionWithPatch")
+
+	var conditionAc *machinev1applyconfigs.ConditionApplyConfiguration
+
+	var statusAc *machinev1applyconfigs.MachineSetStatusApplyConfiguration
+
+	conditionAc = machinev1applyconfigs.Condition().
+		WithType(consts.SynchronizedCondition).
+		WithStatus(status).
+		WithReason(reason).
+		WithMessage(message).
+		WithSeverity(machinev1beta1.ConditionSeverityNone)
+
+	statusAc = machinev1applyconfigs.MachineSetStatus()
+
+	if status == corev1.ConditionTrue {
+		setLastTransitionTime(consts.SynchronizedCondition, mapiMachineSet.Status.Conditions, conditionAc)
+		statusAc.WithSynchronizedGeneration(generation[0])
+	} else {
+		conditionAc.WithSeverity(machinev1beta1.ConditionSeverityError)
+		setLastTransitionTime(consts.SynchronizedCondition, mapiMachineSet.Status.Conditions, conditionAc)
+	}
+
+	// We need to add the conditions to the statusAc
+	statusAc.WithConditions(conditionAc)
+
+	msAc := machinev1applyconfigs.MachineSet(mapiMachineSet.GetName(), mapiMachineSet.GetNamespace()).WithStatus(statusAc)
+	if err := r.Status().Patch(ctx, mapiMachineSet, util.ApplyConfigPatch(msAc), client.ForceOwnership, client.FieldOwner("machineset-sync-controller")); err != nil {
```

Comment:
Why does this failure not get bubbled up? Wouldn't we want to retry if there was an error here?

Reply:
Returning an aggregate would be good yes, but not necessary, you already log the previous error as far as I can tell. Have a see how hard working an aggregate in might be

Reply:
I don't think it's too bad - have done: 

```go
		mapiMachineSet.Spec = newMapiMachineSet.Spec

		if err := r.Update(ctx, mapiMachineSet); err != nil {
			logger.Error(err, "Failed to update MAPI machine set")

			updateErr := fmt.Errorf("failed to update MAPI machine set: %w", err)

			if condErr := r.updateSynchronizedConditionWithPatch(
				ctx, mapiMachineSet, corev1.ConditionFalse,
				"FailedToUpdateMAPIMachineSet", err.Error(), nil); condErr != nil {
				return ctrl.Result{}, utilerrors.NewAggregate([]error{updateErr, condErr})
			}

```